<a href="https://colab.research.google.com/github/inkaromarbekova/python-project/blob/main/Project3_ipynb_Covid_19_Vaccinations_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Capstone 3. Project in Python.**

Hi there! Today I'm going to work on my 3rd project and it will be in Python.The datasets will be downloaded from Kaggle into Google Colab. I'm going to analyze the data regarding the vaccinations in each country and vaccinations by manufacturer per population in each country.Let's get started.

First I'm going to install kaggle.

In [ ]:
# install kaggle version 1.5.12
! pip install --upgrade --force-reinstall --no-deps kaggle

  Using cached kaggle-1.5.12-py3-none-any.whl
  Attempting uninstall: kaggle
    Found existing installation: kaggle 1.5.12
    Uninstalling kaggle-1.5.12:
      Successfully uninstalled kaggle-1.5.12


In [ ]:
# upload kaggle.json file from local system
from google.colab import files
files.upload() 

Saving kaggle.json to kaggle (1).json


{'kaggle.json': b'{"username":"inkaromarbekova","key":"713457924d442da03d4171ac422b61b2"}'}

In [ ]:
# copy the kaggle.json to folder created
!cp kaggle.json ~/.kaggle/

In [ ]:
# permission for the json to act
! chmod 600 /root/.kaggle/kaggle.json

In [ ]:
# list all datasets in kaggle
!kaggle datasets list

ref                                                         title                                              size  lastUpdated          downloadCount  voteCount  usabilityRating  
----------------------------------------------------------  ------------------------------------------------  -----  -------------------  -------------  ---------  ---------------  
gpreda/reddit-vaccine-myths                                 Reddit Vaccine Myths                              233KB  2021-08-22 13:49:51          11105        911  1.0              
crowww/a-large-scale-fish-dataset                           A Large Scale Fish Dataset                          3GB  2021-04-28 17:03:01           6788        509  0.9375           
imsparsh/musicnet-dataset                                   MusicNet Dataset                                   22GB  2021-02-18 14:12:19           2619        191  1.0              
promptcloud/careerbuilder-job-listing-2020                  Careerbuilder Job Listing 2020

The second, I will download the datasets from Kaggle.

In [ ]:
# downloading the 1st dataset from kaggle
!kaggle datasets download -d gpreda/covid-world-vaccination-progress

  0% 0.00/915k [00:00<?, ?B/s]
100% 915k/915k [00:00<00:00, 113MB/s]


In [ ]:
# see the dataset and format
!ls

 covid-world-vaccination-progress.zip   kaggle.json
'kaggle (1).json'		        sample_data


I have to unzip the first dataset because I'm planning to work with csv format.

In [ ]:
# unzip zip file
!unzip \*.zip && rm *.zip

Archive:  covid-world-vaccination-progress.zip
  inflating: country_vaccinations.csv  
  inflating: country_vaccinations_by_manufacturer.csv  


In [ ]:
# see the dataset format changed
!ls

 country_vaccinations_by_manufacturer.csv  'kaggle (1).json'   sample_data
 country_vaccinations.csv		    kaggle.json


In [ ]:
# downloading the 2nd dataset from Kaggle
!kaggle datasets download -d rsrishav/world-population

  0% 0.00/121k [00:00<?, ?B/s]
100% 121k/121k [00:00<00:00, 42.9MB/s]


Here, the same unzipping the second dataset so I can work with csv file.

In [ ]:
# unzip zip file
!unzip \*.zip && rm *.zip

Archive:  world-population.zip
  inflating: 2021_population.csv     
  inflating: timeseries_population_count.csv  


In [ ]:
!ls

 2021_population.csv			    kaggle.json
 country_vaccinations_by_manufacturer.csv   sample_data
 country_vaccinations.csv		    timeseries_population_count.csv
'kaggle (1).json'


**1.1 Import Libraries.**


In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
from scipy import stats

**1.2 Reading Files.**

In [ ]:
# Read the country_vaccinations dataset
vcn = pd.read_csv("country_vaccinations.csv")

In [ ]:
# Read the country_vaccinations_by_manufacturer dataset
vbm = pd.read_csv("country_vaccinations_by_manufacturer.csv")

In [ ]:
# Read the 2021_population dataset
wp = pd.read_csv("2021_population.csv")

**2.1 Country_vaccinations dataset:**

In [ ]:
# Call out the first 5 rows of the dataset
vcn.head()

country iso_code  ...                source_name            source_website
0  Afghanistan      AFG  ...  World Health Organization  https://covid19.who.int/
1  Afghanistan      AFG  ...  World Health Organization  https://covid19.who.int/
2  Afghanistan      AFG  ...  World Health Organization  https://covid19.who.int/
3  Afghanistan      AFG  ...  World Health Organization  https://covid19.who.int/
4  Afghanistan      AFG  ...  World Health Organization  https://covid19.who.int/

[5 rows x 15 columns]

In [ ]:
# Dataset information
vcn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39113 entries, 0 to 39112
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   country                              39113 non-null  object 
 1   iso_code                             39113 non-null  object 
 2   date                                 39113 non-null  object 
 3   total_vaccinations                   21551 non-null  float64
 4   people_vaccinated                    20604 non-null  float64
 5   people_fully_vaccinated              17728 non-null  float64
 6   daily_vaccinations_raw               17713 non-null  float64
 7   daily_vaccinations                   38855 non-null  float64
 8   total_vaccinations_per_hundred       21551 non-null  float64
 9   people_vaccinated_per_hundred        20604 non-null  float64
 10  people_fully_vaccinated_per_hundred  17728 non-null  float64
 11  daily_vaccinations_per_milli

The dataset looks clean.

In [ ]:
vcn['country'].nunique()

222

There are 222 countries listed in the country_vaccinations dataset.

How many rows, columns?

In [ ]:
vcn.shape

(39113, 15)

Showing the datatypes.

In [ ]:
vcn.dtypes

country                                 object
iso_code                                object
date                                    object
total_vaccinations                     float64
people_vaccinated                      float64
people_fully_vaccinated                float64
daily_vaccinations_raw                 float64
daily_vaccinations                     float64
total_vaccinations_per_hundred         float64
people_vaccinated_per_hundred          float64
people_fully_vaccinated_per_hundred    float64
daily_vaccinations_per_million         float64
vaccines                                object
source_name                             object
source_website                          object
dtype: object

In [ ]:
vcn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39113 entries, 0 to 39112
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   country                              39113 non-null  object 
 1   iso_code                             39113 non-null  object 
 2   date                                 39113 non-null  object 
 3   total_vaccinations                   21551 non-null  float64
 4   people_vaccinated                    20604 non-null  float64
 5   people_fully_vaccinated              17728 non-null  float64
 6   daily_vaccinations_raw               17713 non-null  float64
 7   daily_vaccinations                   38855 non-null  float64
 8   total_vaccinations_per_hundred       21551 non-null  float64
 9   people_vaccinated_per_hundred        20604 non-null  float64
 10  people_fully_vaccinated_per_hundred  17728 non-null  float64
 11  daily_vaccinations_per_milli

**I have to use drop formula so to do PearsonR Correlation.**

In [ ]:
vcn = vcn.dropna()
vcn = vcn.reset_index()
vcn

index  ...                                     source_website
0         94  ...                           https://covid19.who.int/
1        101  ...                           https://covid19.who.int/
2        210  ...  https://shendetesia.gov.al/vaksinimi-anticovid...
3        292  ...  https://shendetesia.gov.al/vaksinimi-anticovid...
4        293  ...  https://shendetesia.gov.al/vaksinimi-anticovid...
...      ...  ...                                                ...
14984  39108  ...  https://www.arcgis.com/home/webmap/viewer.html...
14985  39109  ...  https://www.arcgis.com/home/webmap/viewer.html...
14986  39110  ...  https://www.arcgis.com/home/webmap/viewer.html...
14987  39111  ...  https://www.arcgis.com/home/webmap/viewer.html...
14988  39112  ...  https://www.arcgis.com/home/webmap/viewer.html...

[14989 rows x 16 columns]

It looks great, now we can put our formula to see PearsonR correlation between people who got 1st shot and who is fully vaccinated.

In [ ]:
stats.pearsonr(vcn['people_vaccinated'], vcn['people_fully_vaccinated'])

(0.862915278007854, 0.0)

As you see it shows, the correlation coefficient of two variables is zero, there is no linear relationship between the variables. However, this is only for a linear relationship. 

We did statistics , because the data shows COVID-19 infections among fully vaccinated people are low. Also it helps to seek the gap between the 1st shot and the second one, for example in Canada it was hard to get the second one. Also the hospitalization and death rates after getting 1st vaccine.

**2.2 Country_vaccinations_by_manufacturers:**

In [ ]:
# Call the first 5 rows of the dataset
vbm.head()

location        date             vaccine  total_vaccinations
0  Austria  2021-01-08     Johnson&Johnson                   0
1  Austria  2021-01-08             Moderna                   0
2  Austria  2021-01-08  Oxford/AstraZeneca                   0
3  Austria  2021-01-08     Pfizer/BioNTech               31096
4  Austria  2021-01-15     Johnson&Johnson                   0

In [ ]:
# Dataset information.
vbm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12477 entries, 0 to 12476
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   location            12477 non-null  object
 1   date                12477 non-null  object
 2   vaccine             12477 non-null  object
 3   total_vaccinations  12477 non-null  int64 
dtypes: int64(1), object(3)
memory usage: 390.0+ KB


The dataset looks good.

In [ ]:
vbm['location'].nunique()

35

In the country_vaccinations_by_manufacturers, there are only 35 countries listed.

How many rows and columns?

In [ ]:
vbm.shape

(12477, 4)

Datatypes of this file.

In [ ]:
vbm.dtypes

location              object
date                  object
vaccine               object
total_vaccinations     int64
dtype: object

2.3. World_population dataset.

In [ ]:
# Call out the first 5 rows of the dataset.
wp.head()

iso_code        country 2021_last_updated  ... growth_rate world_% rank
0      CHN          China     1,444,936,751  ...       0.34%  18.34%    1
1      IND          India     1,395,407,543  ...       0.97%  17.69%    2
2      USA  United States       333,204,257  ...       0.58%   4.23%    3
3      IDN      Indonesia       276,785,556  ...       1.04%   3.51%    4
4      PAK       Pakistan       225,838,541  ...       1.95%   2.86%    5

[5 rows x 9 columns]

In [ ]:
# Dataset information
wp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 228 entries, 0 to 227
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   iso_code           228 non-null    object
 1   country            228 non-null    object
 2   2021_last_updated  228 non-null    object
 3   2020_population    228 non-null    object
 4   area               228 non-null    object
 5   density_sq_km      228 non-null    object
 6   growth_rate        228 non-null    object
 7   world_%            228 non-null    object
 8   rank               228 non-null    int64 
dtypes: int64(1), object(8)
memory usage: 16.2+ KB


The dataset is perfect too.

In [ ]:
wp['country'].nunique()

228

There are 228 countries in world population file.

How many rows and columns?

In [ ]:
wp.shape

(228, 9)

The datatypes of world population file.

In [ ]:
wp.dtypes

iso_code             object
country              object
2021_last_updated    object
2020_population      object
area                 object
density_sq_km        object
growth_rate          object
world_%              object
rank                  int64
dtype: object

There are 228 countries listed in this dataset.

**2.4 Concatenate two dataframes:**

To estimate the rate of fully vaccinated people of one country per its population, I needed to concat two dataframes. As we can see, in the world population dataset, there are 228 countries listed while in country vaccinations dataset there are only 222 countries. That means we need to clean and remove countries that are mentioned once.

 The idea here is to choose the neccesary columns, concat the chosen dataframes, filter out values that appear more than once in the new dataframes (remove values that appear only in 1 of the 2 dataframes which means that they are redundant data), drop duplicated values (keep last and first in order to keep the neccesary data) and reset their index. 

In [ ]:
wp_sort = wp[['country', '2021_last_updated']] 
wp_sort.sort_values('country') 
vcn_drop = vcn.drop_duplicates('country', keep = "last") 
vcn_sort = vcn_drop[['country', 'people_fully_vaccinated']]
df_same = pd.concat([wp_sort,vcn_sort]) 
df_same = df_same[df_same.groupby('country').country.transform(len) > 1] 
df_same = df_same.drop_duplicates('country', keep = "last") 
df_same_sort = df_same[['country', 'people_fully_vaccinated']] 
df_same_sort = df_same_sort.rename(columns={'country' : 'country_vaccinations'}) 
df_same_sort.reset_index(drop=True, inplace=True) 
wp_clean = pd.concat([wp, df_same])
wp_clean = wp_clean[wp_clean.groupby('country').country.transform(len) > 1]
wp_clean = wp_clean.drop_duplicates('country', keep = "first")
wp_clean_sort = wp_clean[['country', '2021_last_updated']]
wp_clean_sort = wp_clean_sort.sort_values('country')
wp_clean_sort.reset_index(drop = True, inplace=True)
cbn = pd.concat([wp_clean_sort, df_same_sort], axis=1)
cbn

country  ... people_fully_vaccinated
0      Afghanistan  ...                148505.0
1          Albania  ...                575342.0
2          Andorra  ...                  4484.0
3        Argentina  ...              11479360.0
4            Aruba  ...                 68498.0
..             ...  ...                     ...
133  United States  ...             170406785.0
134        Uruguay  ...               2466457.0
135        Vietnam  ...               1636372.0
136         Zambia  ...                227875.0
137       Zimbabwe  ...               1451901.0

[138 rows x 4 columns]

We already had our wanted dataframe which is the combination of the country_vaccinations dataframe and the world_population dataframe with no unwanted country. Let's check once again to see whether the new dataframe is really clean or not.

In [ ]:
cbn.isna().sum()

country                    0
2021_last_updated          0
country_vaccinations       0
people_fully_vaccinated    0
dtype: int64

As we can see, there are 19 null values in the people_fully_vaccinated column. Therefore, I needed to drop all of it.

In [ ]:
cbn = cbn.dropna()
cbn = cbn.reset_index()
cbn

level_0  index  ... country_vaccinations people_fully_vaccinated
0          0      0  ...          Afghanistan                148505.0
1          1      1  ...              Albania                575342.0
2          2      2  ...              Andorra                  4484.0
3          3      3  ...            Argentina              11479360.0
4          4      4  ...                Aruba                 68498.0
..       ...    ...  ...                  ...                     ...
133      133    133  ...        United States             170406785.0
134      134    134  ...              Uruguay               2466457.0
135      135    135  ...              Vietnam               1636372.0
136      136    136  ...               Zambia                227875.0
137      137    137  ...             Zimbabwe               1451901.0

[138 rows x 6 columns]

In [ ]:
cbn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 138 entries, 0 to 137
Data columns (total 5 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   index                    138 non-null    int64  
 1   country                  138 non-null    object 
 2   2021_last_updated        138 non-null    object 
 3   country_vaccinations     138 non-null    object 
 4   people_fully_vaccinated  138 non-null    float64
dtypes: float64(1), int64(1), object(3)
memory usage: 5.5+ KB


Here, I wanted to show the PearsonR correlation between population and the people who are vaccinated. So I have to change the population (string) into float.

In [ ]:
cbn['2021_last_updated'] = cbn['2021_last_updated'].astype(float)

In [ ]:
# checking the numbers under population column
cbn.head()

index      country  ... country_vaccinations people_fully_vaccinated
0      0  Afghanistan  ...          Afghanistan                148505.0
1      1      Albania  ...              Albania                575342.0
2      2      Andorra  ...              Andorra                  4484.0
3      3    Argentina  ...            Argentina              11479360.0
4      4        Aruba  ...                Aruba                 68498.0

[5 rows x 5 columns]

In [ ]:
stats.pearsonr(cbn['2021_last_updated'], cbn['people_fully_vaccinated'])

(0.8442561326578909, 2.1822235742349781e-38)

As you see greater than 0, positive correlation between the overall population and people who are fully vaccinated.

We test this data , to see how many people are getting fully vaccinated in each country comparing with overall population, this can be used to see the situation in each country, and divide it into developed, developing and the 3rd world countries. Predict when we will finish the vaccinations all over the world. Also from this data scientists can see why the variances of Covid 19 is appearing, when the whole world is not vaccinated at the same time , at least in a same period of time virus can mutate, and now maybe we need the 3rd shot.

Let's continue what we were doing.When cleaning data is to check whether that targeted data is in the right type or not. In further calculation, I worked with values from 2021_last_updated and people_fully_vaccinated. Therefore, I had to make sure that they were in the same type.

In [ ]:
# Check data types
cbn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 138 entries, 0 to 137
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   level_0                  138 non-null    int64  
 1   index                    138 non-null    int64  
 2   country                  138 non-null    object 
 3   2021_last_updated        138 non-null    object 
 4   country_vaccinations     138 non-null    object 
 5   people_fully_vaccinated  138 non-null    float64
dtypes: float64(1), int64(2), object(3)
memory usage: 6.6+ KB


In [ ]:
# Reformat data type
cbn.replace(',','', regex=True, inplace=True)
cbn['2021_last_updated'] = cbn['2021_last_updated'].map(lambda x: float(x))
cbn['people_fully_vaccinated'] = cbn['people_fully_vaccinated'].map(lambda x: float(x))
cbn

level_0  index  ... country_vaccinations  people_fully_vaccinated
0          0      0  ...          Afghanistan                 148505.0
1          1      1  ...              Albania                 575342.0
2          2      2  ...              Andorra                   4484.0
3          3      3  ...            Argentina               11479360.0
4          4      4  ...                Aruba                  68498.0
..       ...    ...  ...                  ...                      ...
133      133    133  ...        United States              170406785.0
134      134    134  ...              Uruguay                2466457.0
135      135    135  ...              Vietnam                1636372.0
136      136    136  ...               Zambia                 227875.0
137      137    137  ...             Zimbabwe                1451901.0

[138 rows x 6 columns]

 I had to clean data once again to make sure that there was no weird value or logic. In this case, I would check whether there was any country with the number of fully vaccinated people was higher than the last updated population number.

In [ ]:
# Check weird data
cbn[cbn['2021_last_updated'] < cbn['people_fully_vaccinated']]

level_0  index  ... country_vaccinations  people_fully_vaccinated
45       45     45  ...            Gibraltar                  39162.0

[1 rows x 6 columns]

In [ ]:
# Remove weird data
cbn = cbn.drop(index = [61])

**3.1 Exploratory Analysis**

Let's analyze the global country vaccinations progress.

1.What are the top 5 biggest countries with 
vaccinations progress ?
2.What are the top 5 smallest countries with vaccinations process?
3.What vaccine is most common and least used ?
4.What is the global average vaccinations by month ?
5.What are the top 5 countries that are having the highest and lowest of fully vaccinated people per population ?

**What are the top 5 biggest countries with vaccinations progress ?**

In [ ]:
# Five Highest Countries
total = vcn.groupby('country')['total_vaccinations'].max().reset_index()
fhc = total.sort_values('total_vaccinations', ascending=False).head(5)
pd.set_option('display.float_format', lambda x: '%.0f'% x)
fhc

country  total_vaccinations
27           China           845299000
63           India           576365820
147  United States           361684564
19          Brazil           173650317
70           Japan           115739998

Here, you can see that China, known as the first country that the COVID-19 pandemic occurred has the highest number of total vaccinations. Nearly 1.92 billions vaccine doses were used. It is reasonable because China has faced the pandemic for 2 years and they population is the highest in the world. The next 3 positions are for India, United States and Brazil. It is expected because these 3 countries have the highest number of cases.

In [ ]:
fig = px.bar(fhc, 
             x='country', 
             y='total_vaccinations',
             labels = {'country' : 'Country', 'total_vaccinations' : 'Total Vaccinations'},
             title = "Top 5 Countries With Biggest Vaccinations Progress"
            )
fig.show()

 China is dominating in the number of vaccine used by this graph.

 **What are the top 5 smallest countries with vaccinations progress ?**

In [ ]:
# Top 5 smallest countries with vaccinations progress
flc = total.sort_values('total_vaccinations', ascending=False).tail(5)
flc

country  total_vaccinations
47             Gabon               33591
58             Haiti               19182
2            Andorra               15269
54           Grenada               13458
46  French Polynesia               10097

It's weird in Pitcairn 83 people is vaccinated but ion fact the population here is 50 people.

In [ ]:
fig = px.bar(flc, 
             x='country', 
             y='total_vaccinations',
             labels = {'country' : 'Country', 'total_vaccinations' : 'Total Vaccinations'},
             title = "Top 5 Countries With Lowest Vaccinations Progress"
            )
fig.show()

**What vaccine is most common used and least common used ?**

Let's explore the popularity of vaccines by summing their total up. In this case, like the total_vaccinations column in the country_vaccinations dataset, the value in the total_vaccinations in country_vaccinations_by_manufacturers are cummulative. Therefore, I used max function to take out the lastest value of each vaccine.

In [ ]:
# Call out the number of total vaccines used by country and vaccine types
vpc = vbm.groupby(['vaccine', 'location'])['total_vaccinations'].max().reset_index()
vpc

vaccine   location  total_vaccinations
0            CanSino      Chile              515836
1    Johnson&Johnson    Austria              230542
2    Johnson&Johnson    Belgium              337333
3    Johnson&Johnson   Bulgaria              112434
4    Johnson&Johnson    Croatia               59920
..               ...        ...                 ...
127          Sinovac      Chile            19049522
128          Sinovac  Hong Kong             2645923
129          Sinovac    Uruguay             3242607
130        Sputnik V    Hungary             1798288
131        Sputnik V   Slovakia               37947

[132 rows x 3 columns]

In [ ]:
# Rank the popularity of vaccines
vr = vpc.groupby('vaccine')['total_vaccinations'].sum().reset_index()
vr = vr.sort_values('total_vaccinations', ascending=False)
vr

vaccine  total_vaccinations
4     Pfizer/BioNTech           686904224
2             Moderna           206941323
3  Oxford/AstraZeneca            83312521
1     Johnson&Johnson            28253742
6             Sinovac            24938052
5   Sinopharm/Beijing             2081012
7           Sputnik V             1836235
0             CanSino              515836

In [ ]:
# Pieplot
fig = px.pie(vr, values='total_vaccinations', names='vaccine', title='Vaccines Occupancy' )
fig.show()

This pie illustrates the real domination of Pfizer/BioNTech.

In [ ]:
# Barplot
fig = px.bar(vr, 
             y='total_vaccinations', 
             x='vaccine',
             labels = {'vaccine' : 'Vaccines', 'total_vaccinations' : 'Total Vaccinations'},
             title = "Overview of Vaccines"
            )
fig.show()

There are only 33 countries mentioned in the dataset. However, it is enough to figure out that Pfizer/BioNTech and Moderna are the leaders in this race. While, the two vaccines made by Chinese (CanSino, Sinopharm/Beijing) and one vaccine made by Russia look unpopular. 

Since one country may uses several types of vaccines. I decided to make a table to look at the total number of vaccines they have used so far.

In [ ]:
# Creat pivot table
pvt = pd.pivot_table(data = vbm, index = ['location'], columns = ['vaccine'], values = 'total_vaccinations', aggfunc = 'max')
pvt.fillna(0)

vaccine        CanSino  Johnson&Johnson  ...  Sinovac  Sputnik V
location                                 ...                    
Austria              0           230542  ...        0          0
Belgium              0           337333  ...        0          0
Bulgaria             0           112434  ...        0          0
Chile           515836                0  ... 19049522          0
Croatia              0            59920  ...        0          0
Cyprus               0            19178  ...        0          0
Czechia              0           183951  ...        0          0
Denmark              0            47950  ...        0          0
Estonia              0            43933  ...        0          0
Finland              0                0  ...        0          0
France               0           978897  ...        0          0
Germany              0          2692945  ...        0          0
Hong Kong            0                0  ...  2645923          0
Hungary              0           133015  ...        0    1798288
Iceland              0            53569  ...        0          0
Ireland              0           227896  ...        0          0
Italy                0          1412441  ...        0          0
Japan                0                0  ...        0          0
Latvia               0           110136  ...        0          0
Liechtenstein        0                0  ...        0          0
Lithuania            0           319074  ...        0          0
Luxembourg           0            34139  ...        0          0
Malta                0            19438  ...        0          0
Netherlands          0           782858  ...        0          0
Poland               0          1622194  ...        0          0
Portugal             0          1011475  ...        0          0
Romania              0           556964  ...        0          0
Slovakia             0            53353  ...        0      37947
Slovenia             0            78842  ...        0          0
South Korea          0          1129784  ...        0          0
Spain                0          1902392  ...        0          0
Sweden               0                0  ...        0          0
Switzerland          0                0  ...        0          0
United States        0         14099089  ...        0          0
Uruguay              0                0  ...  3242607          0

[35 rows x 8 columns]

In USA 14050126 of Johnson & Johnson, 142827859 of Moderna, 203416431 of Pfizer/BioNTech doses is used.

In [ ]:
pvt.info()

<class 'pandas.core.frame.DataFrame'>
Index: 35 entries, Austria to Uruguay
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   CanSino             1 non-null      float64
 1   Johnson&Johnson     28 non-null     float64
 2   Moderna             32 non-null     float64
 3   Oxford/AstraZeneca  30 non-null     float64
 4   Pfizer/BioNTech     35 non-null     float64
 5   Sinopharm/Beijing   1 non-null      float64
 6   Sinovac             3 non-null      float64
 7   Sputnik V           2 non-null      float64
dtypes: float64(8)
memory usage: 2.5+ KB


In [ ]:
pvt = pvt.dropna()
pvt = pvt.reset_index()
pvt

Empty DataFrame
Columns: [location, CanSino, Johnson&Johnson, Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm/Beijing, Sinovac, Sputnik V]
Index: []

In [ ]:
pvt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 0 entries
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   location            0 non-null      object 
 1   CanSino             0 non-null      float64
 2   Johnson&Johnson     0 non-null      float64
 3   Moderna             0 non-null      float64
 4   Oxford/AstraZeneca  0 non-null      float64
 5   Pfizer/BioNTech     0 non-null      float64
 6   Sinopharm/Beijing   0 non-null      float64
 7   Sinovac             0 non-null      float64
 8   Sputnik V           0 non-null      float64
dtypes: float64(8), object(1)
memory usage: 124.0+ bytes


**What is the global average vaccinations by month ?**

Let's check the progress of vaccinations through months to see wheter the progress is increasing or not. Here, I calculated the mean of the global daily vaccinations under the reference of month to have the overview.

In [ ]:
# Find the global average total vaccinations by month
vcn['date'] = pd.to_datetime(vcn['date'])
avg = vcn.groupby(vcn['date'].dt.strftime('%B'))['daily_vaccinations'].mean().sort_values().reset_index()
avg

date  daily_vaccinations
0  December                9693
1   January               61968
2  February               90488
3     March              139108
4       May              144728
5     April              160450
6      June              196783
7      July              197303
8    August              200979

If you look closely, you will see that July came before June. Therefore, I would reindex these two months.

In [ ]:
avg = avg.reindex([0, 1, 2, 3, 4, 5, 7, 6, 8])
avg

date  daily_vaccinations
0  December                9693
1   January               61968
2  February               90488
3     March              139108
4       May              144728
5     April              160450
7      July              197303
6      June              196783
8    August              200979

In [ ]:
# Lineplot to see the full progress
fig = px.line(avg, 
             x='date', 
             y='daily_vaccinations',
             labels = {'daily_vaccinations' : 'Global Monthly Vaccinations', 'date' : 'Month'},
             title = "Average Total Vaccinations"
            )
fig.show()


In conclusion, 
The data shows that countries like China and India that have a similar population of 1.4 billion have had a very different success rate when it comes to the implementation of the vaccination program. The gap between fully vaccinated people in these two countries is substantial. This gap can be one of the main contributors to the difference between the two countries death and hospitalization rates. Furthermore, the gap may also be a contributor in the effect the pandemic has had on the economies of the two countries.

As we can see from the data Pfizer(66,4%) and Moderna(20%) are leading the world in vaccination rates. These two vaccines require two doses for full effectiveness.
The data supports the hypothesis that fully vaccinated people are less susceptible to covid then partially vaccinated people, the data also shows that the gap between the two doses may play a role in the effectiveness of the vaccine. Furthermore, other factors such as age and preexisting health conditions seem to play a role as well.

The effectiveness of theses vaccines is contingent on many factors some are changing by the day, for example; new variants of the virus are appearing throughout the world.
More data is needed in order to isolate the many variables and enable us to draw meaningful conclusions.